source/MaterialXTest/ShaderValid.cpp

```diff
@@ -1022,8 +1060,104 @@ TEST_CASE("MaterialX documents", "[shadervalid]")
     // Dump out profiling information
     totalTime.endTimer();
     profileTimes.print(profilingLog);
-    profilingLog << "------------------" << std::endl;
-    options.print(profilingLog);
+
+
+    if (options.checkImplCount)
+    {
+        profilingLog << "---------------------------------------" << std::endl;
+
+        // Get implementation count from libraries
+        std::set<mx::ImplementationPtr> libraryImpls;
+        const std::vector<mx::ElementPtr>& children = dependLib->getChildren();
+        for (auto child : children)
+        {
+            mx::ImplementationPtr impl = child->asA<mx::Implementation>();
+            if (!impl)
+            {
+                continue;
+            }
+            libraryImpls.insert(impl);
+        }
+
+        const std::unordered_map<std::string, mx::ShaderNodeImplPtr>* oslImpls = nullptr;
+#ifdef MATERIALX_BUILD_GEN_GLSL
+        oslImpls = &(oslShaderGenerator->getImplementationsUsed());
+#endif
+        const std::unordered_map<std::string, mx::ShaderNodeImplPtr>* glslImpls = nullptr;
+#ifdef MATERIALX_BUILD_GEN_OSL
+        glslImpls = &(glslShaderGenerator->getImplementationsUsed());
+#endif
+        size_t skipCount = 0;
+        profilingLog << "-- Possibly missed implementations ----" << std::endl;
+        std::vector<std::string> whiteList =
+        {
+            "arrayappend", "backfacing", "screen", "curveadjust", "dot_surfaceshader", "mix_surfaceshader"
+            "displacementShader", "displacementshader", "volumeshader", "IM_dot_filename", "ambientocclusion", "dot_lightshader"
+        };
+        const std::string OSL_STRING("osl");
+        const std::string SX_OSL_STRING("sx_osl");
+        for (auto libraryImpl : libraryImpls)
```

Comment:
For each library impl test:
1. Is it on the whitelist so are allowed to not be tested (since there are no impls at this point
2. Are directly referenced and not cached in shader generators
3. Are used and cached by shader generators
4. Are overridden. For now this only takes into account sx-osl overrides on osl. There is no mechanism to store what impls are overridden at this point and is rather costly to rescan.

Reply:
With the `getCachedImplementation()` on the ShaderGenerator and CMS classes maybe we could directly check if a given implementation element is used, and report if it's not. I think that would be simpler and more robust then counting the number in use and comparing counters. We can then also report which one is untested, and not just how many are missing.

Reply:
It is actually performing a per implementation check as you say. But it still requires a check against non-cached implementations. 
The trivial end check is just to make sure the library count is the same as the usage count.

Reply:
The only alternative is if each shadergen and cms keeps track of all the implementations they "could" use and "check them off" during generation. I don't really like this as it's unncessary overhead for what should be an external test.